relays/ethereum/src/ethereum_deploy_contract.rs

```diff
@@ -0,0 +1,136 @@
+			eth: Default::default(),
+			eth_sign: Default::default(),
+			// compiler: 0.6.6+commit.6c089d02 + optimization
+			eth_contract_code: hex::decode(include_str!("../res/substrate-bridge-bytecode.hex"))
+				.expect("code is hardcoded, thus valid; qed"),
+			sub: Default::default(),
+			sub_initial_authorities_set_id: None,
+			sub_initial_authorities_set: None,
+			sub_initial_header: None,
+		}
+	}
+}
+
+/// Deploy Bridge contract on Ethereum chain.
+pub fn run(params: EthereumDeployContractParams) {
+	let mut local_pool = futures::executor::LocalPool::new();
+
+	let result = local_pool.run_until(async move {
+		let eth_client = ethereum_client::client(params.eth);
+		let sub_client = substrate_client::client(params.sub);
+
+		let (sub_client, initial_header) = prepare_initial_header(sub_client, params.sub_initial_header).await;
+		let (initial_header_hash, initial_header) = initial_header?;
+		let initial_set_id = params.sub_initial_authorities_set_id.unwrap_or(0);
+		let (_, initial_set) = prepare_initial_authorities_set(
+			sub_client,
+			initial_header_hash,
+			params.sub_initial_authorities_set,
+		).await;
+		let initial_set = initial_set?;
+
+		log::info!(
+			target: "bridge",
+			"Deploying Ethereum contract.\r\n\tInitial header: {}\r\n\tInitial authorities set ID: {}\r\n\tInitial authorities set: {}",
+			hex::encode(&initial_header),
+			initial_set_id,
+			hex::encode(&initial_set),
+		);
+
+		ethereum_client::deploy_bridge_contract(
+			eth_client,
+			&params.eth_sign,
+			params.eth_contract_code,
+			initial_header,
+			initial_set_id,
+			initial_set,
+		).await.1.map_err(|error| format!("Error deploying contract: {:?}", error))
+	});
+
+	if let Err(error) = result {
+		log::error!(target: "bridge", "{}", error);
+	}
+}
+
+/// Prepare initial header.
+async fn prepare_initial_header(
+	sub_client: substrate_client::Client,
+	sub_initial_header: Option<Vec<u8>>,
+) -> (substrate_client::Client, Result<(SubstrateHash, Vec<u8>), String>) {
+	match sub_initial_header {
+		Some(raw_initial_header) => match SubstrateHeader::decode(&mut &raw_initial_header[..]) {
+			Ok(initial_header) => (sub_client, Ok((initial_header.hash(), raw_initial_header))),
+			Err(error) => (sub_client, Err(format!("Error decoding initial header: {}", error))),
+		},
+		None => {
+			let (sub_client, initial_header) = substrate_client::header_by_number(sub_client, Zero::zero()).await;
+			(
+				sub_client,
+				initial_header
+					.map(|header| (header.hash(), header.encode()))
+					.map_err(|error| format!("Error reading Substrate genesis header: {:?}", error)),
+			)
+		}
+	}
+}
+
+/// Prepare initial GRANDPA authorities set.
+async fn prepare_initial_authorities_set(
+	sub_client: substrate_client::Client,
+	sub_initial_header_hash: SubstrateHash,
+	sub_initial_authorities_set: Option<Vec<u8>>,
+) -> (substrate_client::Client, Result<Vec<u8>, String>) {
+	let (sub_client, initial_authorities_set) = match sub_initial_authorities_set {
```

Comment:
The `Client` handling is so cumbersome that I'm even inclined to pass it via `thread_local` or `lazy_static`. Something like this:
```rust
static mut CLIENT: Option<Client> = None;

struct WithClient;
impl WithClient {
    pub fn new(s: Client) -> Self {
        unsafe {
            if CLIENT.is_some() {
                panic!("There should be only one instance of client");
            }
            CLIENT = Some(s);
        }
        WithClient
    }

    pub fn with_client<R>(&mut self, call: impl FnMut(Client) -> (Client, R)) -> R {
        let client = unsafe { CLIENT.take().expect("...") };
        let (client, result) = call(client);
        unsafe { CLIENT = Some(client); }
        result
    }
}
```

You can then keep passing `WithClient` mutable reference (or agregate all functions in a single struct) and return regular `Result<>` object (so `?` would work), but only in places where you really need to own the client you would fall back to this tuple-return format.


Or simply the functions assume the client is initialized in `lazy_static!` and we don't pass any extra objects at all (this will require Mutex though)

Reply:
NVM, the whole thing. Should be reworked as per: https://github.com/paritytech/parity-bridges-common/issues/72